src/librustc_typeck/check/demand.rs

```diff
@@ -374,6 +379,77 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         None
     }
 
+    // This function checks if the specified expression is a built-in range literal
+    // (See: librustc/hir/lowering.rs::LoweringContext::lower_expr() )
+    fn is_range_literal(&self, expr: &hir::Expr) -> bool {
+        use hir::{Path, QPath, ExprKind, TyKind};
+
+        // we support `::std::ops::Range` and `::std::core::Range` prefixes
+        // (via split on "|")
+        let ops_path = ["{{root}}", "std|core", "ops"];
+
+        let is_range_path = |path: &Path| {
+            let ident_names: Vec<_> = path.segments
+                .iter()
+                .map(|seg| seg.ident.as_str())
+                .collect();
+
+            if let Some((last, preceding)) = ident_names.split_last() {
+                last.starts_with("Range") &&
+                    preceding.len() == 3 &&
+                    preceding.iter()
+                        .zip(ops_path.iter())
+                        .all(|(seg, match_seg)| {
+                            match_seg.split("|")
+                                .into_iter()
+                                .any(|ref spl_seg| seg == spl_seg)
+                        })
+            } else {
+                false
+            }
+        };
+
+        let is_range_struct_snippet = |span: &Span| {
```

Comment:
I'd negate this function: call it something like `is_range_literal` and invert the return value. You currently just use it as `!is_range_struct_snippet`, so it makes sense to flip it here.

Reply:
I shall rename, but one question: `is_range_literal` would shadow the encompassing `fn` (same name). Isn't that frowned upon in Rust code?

Reply:
Oh, I didn't spot that: `is_range_path` is clearer, you're right 👍